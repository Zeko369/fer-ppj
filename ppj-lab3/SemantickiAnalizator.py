import sys


class SemanticAnalyzer:
    def __init__(self, input):
        self.input = input
        self.lines = input.splitlines()

        self.global_vars: dict[str, int] = {}
        self.scope_vars: list[dict[str, int]] = []

    def run(self):
        loop = 0

        i = 1
        while i < len(self.lines):
            line = self.lines[i]
            line_type = self.get_type(line)

            if line_type == 'KR_ZA':
                loop += 1
                self.scope_vars.append({})
            elif line_type == 'KR_AZ':
                loop -= 1
                self.scope_vars.pop()
            elif line_type == 'IDN':
                idn = self.get_identifier(line)
                line_number = self.get_line(line)
                next_line_type = self.get_type(self.lines[i + 1])

                if next_line_type == 'OP_PRIDRUZI':
                    if loop == 0 and idn not in self.global_vars:
                        self.global_vars[idn] = line_number
                    if loop > 0 and idn not in self.global_vars and idn not in self.scope_vars[-1]:
                        self.scope_vars[-1][idn] = line_number

                    i += 1
                    continue

                if next_line_type == 'KR_OD':
                    self.scope_vars[-1][idn] = line_number
                    i += 1
                    continue

                def_line_number = self.is_local(idn)
                if def_line_number is None:
                    def_line_number = self.global_vars.get(idn)

                if def_line_number == line_number:
                    def_line_number = None

                if def_line_number is None:
                    print('err', line_number, idn)
                    break

                print(line_number, def_line_number, idn)

            i += 1

    def is_local(self, idn: str):
        for scope in self.scope_vars:
            if idn in scope:
                return scope[idn]

    def get_line(self, line: str):
        return int(line.strip().split(' ')[1])

    def get_type(self, line: str):
        return line.strip().split(' ')[0]

    def get_identifier(self, line: str):
        if self.get_type(line) != 'IDN':
            raise Exception('Not an identifier')

        return line.strip().split(' ')[2]


if __name__ == '__main__':
    SemanticAnalyzer(sys.stdin.read()).run()
