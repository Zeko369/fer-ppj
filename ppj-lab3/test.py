from contextlib import redirect_stdout
from io import StringIO
import os

from SemantickiAnalizator import SemanticAnalyzer


def load_file(filename):
    with open(filename, 'r') as f:
        return f.read()


def load_test(test):
    input_string = load_file(os.path.join('./tests', test, 'test.in'))
    expected_output = load_file(os.path.join('./tests', test, 'test.out'))

    return input_string, expected_output


class MyStringIO(StringIO):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.output = ''

    def write(self, s):
        self.output += s


if __name__ == '__main__':
    tests = sorted(os.listdir('./tests'))
    # tests = ['15_gen']
    score = 0

    for test in tests:
        input_string, expected_output = load_test(test)

        io = MyStringIO()
        try:
            with redirect_stdout(io):
                SemanticAnalyzer(input_string).run()
        except Exception as e:
            print(f"Run: ./tests/{test}/test.out failed")
            continue

        if io.output == expected_output:
            score += 1
        else:
            out = f'./tests/{test}/test.out'
            actual = f'./tests/{test}/test.actual'

            msg = f"Test: {out} failed, output was {actual}"
            print(msg)

            print(io.output)

            with open(os.path.join('./tests', test, 'test.actual'), 'w') as f:
                f.write(io.output)

    print(f'Score: {score}/{len(tests)}')
