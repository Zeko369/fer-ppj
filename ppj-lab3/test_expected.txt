3 1 n
3 1 n
4 2 rez
4 3 i
4 3 i
4 3 i